src/app/services/shell.service.ts

```diff
@@ -1,78 +1,104 @@
 import {
   Injectable, EventEmitter, Inject,
 } from '@angular/core';
+import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';
 import { environment } from 'environments/environment';
+import { Observable, tap } from 'rxjs';
+import { webSocket as rxjsWebsocket, WebSocketSubject } from 'rxjs/webSocket';
+import { IncomingApiMessageType } from 'app/enums/api-message-type.enum';
+import { WEBSOCKET } from 'app/helpers/websocket.helper';
 import { WINDOW } from 'app/helpers/window.helper';
 import { ShellConnectedEvent } from 'app/interfaces/shell.interface';
+import { TerminalConnectionData } from 'app/interfaces/terminal.interface';
 
+@UntilDestroy()
 @Injectable()
 export class ShellService {
-  pendingMessages: string[] = [];
-  socket: WebSocket;
-  connected = false;
-  token: string;
-  vmId: number;
-  podInfo: {
-    chart_release_name: string;
-    pod_name: string;
-    container_name: string;
-    command: string;
-  };
-
-  private shellCmdOutput: ArrayBuffer;
-  shellOutput = new EventEmitter<ArrayBuffer>();
-  shellConnected = new EventEmitter<ShellConnectedEvent>();
+  private encoder = new TextEncoder();
+  private ws$: WebSocketSubject<unknown>;
+  private connectionUrl = (this.window.location.protocol === 'https:' ? 'wss://' : 'ws://') + environment.remote + '/websocket/shell/';
+  private isConnected = false;
+
+  private token: string;
+  private connectionData: TerminalConnectionData;
+
+  private shellOutput = new EventEmitter<ArrayBuffer>();
+  private shellConnected = new EventEmitter<ShellConnectedEvent>();
+
+  get shellOutput$(): Observable<ArrayBuffer> {
```

Comment:
Even though it was my suggestion initially to make these getters, since we're only going to initialize the `shellOutput` and `shellConnected` EventEmitter instances once, we can use a readonly variable for the observable value instead. As in getters, every time the value is re-evaluated, the `asObservable` method is called again fetching a new observable. Something like 
`readonly shellOutput$ = this.shellOutput.asObservable();`